exporters/otlp/otlp.go

```diff
@@ -37,63 +34,67 @@ import (
 
 type Exporter struct {
 	// mu protects the non-atomic and non-channel variables
-	mu sync.RWMutex
-	// senderMu protects the concurrent unsafe sends on the shared gRPC client connection.
-	senderMu          sync.Mutex
-	started           bool
-	traceExporter     coltracepb.TraceServiceClient
-	metricExporter    colmetricpb.MetricsServiceClient
-	grpcClientConn    *grpc.ClientConn
-	lastConnectErrPtr unsafe.Pointer
-
-	startOnce      sync.Once
-	stopCh         chan bool
-	disconnectedCh chan bool
-
-	backgroundConnectionDoneCh chan bool
-
-	c        config
-	metadata metadata.MD
+	mu       sync.RWMutex
+	senderMu sync.RWMutex
+	started  bool
+
+	metricsConnection *otlpConnection
+	metricsClient     colmetricpb.MetricsServiceClient
+
+	tracesConnection *otlpConnection
+	tracesClient     coltracepb.TraceServiceClient
+
+	startOnce sync.Once
+	stopCh    chan bool
 }
 
 var _ tracesdk.SpanExporter = (*Exporter)(nil)
 var _ metricsdk.Exporter = (*Exporter)(nil)
 
-// newConfig initializes a config struct with default values and applies
-// any ExporterOptions provided.
-func newConfig(opts ...ExporterOption) config {
-	cfg := config{
-		numWorkers:        DefaultNumWorkers,
-		grpcServiceConfig: DefaultGRPCServiceConfig,
-	}
-	for _, opt := range opts {
-		opt(&cfg)
-	}
-	return cfg
-}
-
 // NewExporter constructs a new Exporter and starts it.
-func NewExporter(opts ...ExporterOption) (*Exporter, error) {
-	exp := NewUnstartedExporter(opts...)
+func NewExporter(configuration ConnConfigurations) (*Exporter, error) {
+	exp := NewUnstartedExporter(configuration)
 	if err := exp.Start(); err != nil {
 		return nil, err
 	}
 	return exp, nil
 }
 
 // NewUnstartedExporter constructs a new Exporter and does not start it.
-func NewUnstartedExporter(opts ...ExporterOption) *Exporter {
+func NewUnstartedExporter(configuration ConnConfigurations) *Exporter {
 	e := new(Exporter)
-	e.c = newConfig(opts...)
-	if len(e.c.headers) > 0 {
-		e.metadata = metadata.New(e.c.headers)
+
+	// Only create connections for the signals with configured addresses.
+	if configuration.metrics.collectorAddr != "" {
```

Comment:
A connection for a signal is only created if an address has been given for that signal. Otherwise it's considered the client does not want to use that signal, so no connection is needed.